### PR TITLE
feat: implement metrics gathering for core

### PR DESCRIPTION
**Pull request by sighphyre**
This adds metrics gathering for the core Yggdrasil client, using atomic primitives behind the scenes, which means this is should still be Send & Sync (yay!). The tradeoff here is that the client won't hold metrics longer than the engine holds the internal state - when passing in a new state for compilation, the engine will return unhandled metrics, that's now the callers responsibility to handle unsent metrics

### CODE REVIEW

File: unleash-yggdrasil/src/lib.rs [hunk 1]
```diff
@@ -28,7 +32,47 @@ pub struct CompiledToggle {
...
+    pub variants: Vec<CompiledVariant>,
+    pub yes: AtomicU32,
+    pub no: AtomicU32,
+    pub default_variant: AtomicU32,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct CompiledVariant {
+    pub name: String,
+    pub weight: i32,
```
Comment (nunogois) [inline unleash-yggdrasil/src/lib.rs:44]: I'm curious about this type. Can weight be negative and is it only int?

File: unleash-yggdrasil/src/lib.rs [hunk 2]
```diff
@@ -94,84 +158,144 @@ impl EngineState {
...
+            })
+            .unwrap_or_default()
+    }
+
+    fn resolve_variant<'a>(
+        &self,
+        toggle: &CompiledToggle,
+        variants: &'a Vec<CompiledVariant>,
+        context: &Context,
+    ) -> Option<&'a CompiledVariant> {
```
Comment (nunogois) [inline unleash-yggdrasil/src/lib.rs:231]: From a quick search does `'a` here mean that our result type only lives as long as what we're taking in as our parameters? Would love a ELI5 in this context if possible, thanks!

File: unleash-yggdrasil/src/lib.rs [hunk 3]
```diff
@@ -94,84 +158,144 @@ impl EngineState {
...
+        if variants.is_empty() {
+            return None;
+        }
+        if let Some(found_override) = check_for_variant_override(variants, context) {
+            return Some(found_override);
+        }
+        let total_weight: u32 = variants.iter().map(|var| var.weight as u32).sum();
```
Comment (nunogois) [inline unleash-yggdrasil/src/lib.rs:238]: Seems like we're `as u32`ing the weights here. Related to https://github.com/Unleash/yggdrasil/pull/25/files#r1104638297?

**Review by nunogois — APPROVED** [review]
LGTM

File: unleash-yggdrasil/src/lib.rs (hunk 1 above)
Comment (sighphyre) [reply unleash-yggdrasil/src/lib.rs:44]: So this is Yggdrasil being sandwiched between two competing types. Unleash-types passes weight as an i32 and our hashing library requires a u32. To use weight there needs to be a cast somewhere. I believe the reason we set the weight to an i32 in Unleash-types is because we trust the rigor of our data validation in the typescript version as far as we can throw it and it's better to parse the wrong type and cast it than it is to fail the parse even though you have the data you need

File: unleash-yggdrasil/src/lib.rs (hunk 2 above)
Comment (sighphyre) [reply unleash-yggdrasil/src/lib.rs:231]: Your Google-Fu is most excellent. Right let's unpack this beast. 

In our function signature we declare `fn resolve_variant<'a>` which tells Rust that this function is going to receive some borrowed data and that the return value is _also_ going to be a borrow but it'll be borrowed against the incoming data. The 'a here is telling Rust that we're going to call this lifetime `'a`.

In the parameters we declare `variants: &'a Vec<CompiledVariant>` which tells Rust that this parameter has a lifetime of `'a` - this is the parameter that fulfils the lifetime declaration in the function signature and tells Rust that the lifetime of `'a` is as long as variants lives.

In the return value we have `Option<&'a CompiledVariant>` so now we're telling Rust that the return value must live at least as long as whatever data is bound to the `'a` lifetime.

So in English: we're borrowing a chunk of data called variants, when we return the data from this function, the data in variants can't be discarded before the return value is.

Comment (nunogois) [reply unleash-yggdrasil/src/lib.rs:231]: That's perfect and really helpful, thanks Simon! 
@gastonfournier you may find this insightful as well.
